src/common/buf.c

```diff
@@ -15,6 +16,39 @@ strip_curly_braces(char *s)
 	s[len] = 0;
 }
 
+static void
+buf_add_one_char(struct buf *s, char ch)
+{
+	if (s->alloc <= s->len + 1) {
+		s->alloc = s->alloc * 3 / 2 + 16;
+		s->buf = xrealloc(s->buf, s->alloc);
+	}
+	s->buf[s->len++] = ch;
+	s->buf[s->len] = '\0';
+}
+
+void
+buf_expand_tilde(struct buf *s)
+{
+	struct buf new;
+	buf_init(&new);
+	for (int i = 0 ; i < s->len ; i++) {
+		if (s->buf[i] == '~') {
+			buf_add(&new, getenv("HOME"));
+		} else {
+			buf_add_one_char(&new, s->buf[i]);
+		}
+	}
+	free(s->buf);
+	s->buf = new.buf;
```

Comment:
Should we also copy over new.len` here?

Reply:
Great catch. Done.

Reply:
Wrote a unit test to build confidence around this.
Realized that we need to do the same (copy `new.len` + `new.alloc`) for `buf_expand_shell_variables()`.
It's obvious in hindsight and I don't know why I didn't see it. Amazing what a unit test can do :smile: